pkg/controller/s3/s3bucket.go

```diff
@@ -18,249 +18,216 @@ package s3
 	}
-	bucketInfo, err := client.GetBucketInfo(bucket.Status.IAMUsername, bucket)
+
+	name := meta.GetExternalName(cr)
+
+	// Check if bucket in spec exists
+	_, err := e.client.HeadBucketRequest(&awsS3.HeadBucketInput{
+		Bucket: aws.String(name),
+	}).Send(ctx)
+
 	if err != nil {
-		return r.fail(bucket, err)
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(s3.IsErrorNotFound, err), errGetBucket)
 	}
 
-	if bucketInfo.Versioning != bucket.Spec.Versioning {
-		err := client.UpdateVersioning(bucket)
-		if err != nil {
-			return r.fail(bucket, err)
-		}
-	}
+	policy := e.getBucketPolicy(meta.GetExternalName(cr))
+
+	cr.Status.AtProvider = s3.GenerateObservation(policy)
+
+	cr.Status.SetConditions(runtimev1alpha1.Available())
 
-	// TODO: Detect if the bucket CannedACL has changed, possibly by managing grants list directly.
-	err = client.UpdateBucketACL(bucket)
+	upToDate, err := s3.IsUpToDate(cr.Spec.ForProvider, policy)
 	if err != nil {
-		return r.fail(bucket, err)
+		return managed.ExternalObservation{}, errors.Wrap(err, errUpdate)
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: upToDate,
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*bucketv1beta1.S3Bucket)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
 	}
+	cr.SetConditions(runtimev1alpha1.Creating())
+
+	// Create the bucket
+	_, err := e.client.CreateBucketRequest(s3.GenerateCreateBucketInput(meta.GetExternalName(cr), &cr.Spec.ForProvider)).Send(ctx)
 
-	// Eventually consistent, so we check if this version is newer than our stored version.
-	changed, err := bucket.HasPolicyChanged(bucketInfo.UserPolicyVersion)
 	if err != nil {
-		return r.fail(bucket, err)
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateBucket)
 	}
-	if changed {
-		currentVersion, err := client.UpdatePolicyDocument(bucket.Status.IAMUsername, bucket)
-		if err != nil {
-			return r.fail(bucket, err)
-		}
-		err = bucket.SetUserPolicyVersion(currentVersion)
-		if err != nil {
-			return r.fail(bucket, err)
-		}
+
+	// Attach the policy
+	if cr.Spec.ForProvider.Policy != nil {
+		err = e.putBucketPolicy(cr.Spec.ForProvider.Policy, meta.GetExternalName(cr))
 	}
 
-	bucket.Status.SetConditions(runtimev1alpha1.ReconcileSuccess())
-	return result, r.Update(ctx, bucket)
+	return managed.ExternalCreation{}, errors.Wrap(err, errPolicyAttach)
 }
 
-func (r *Reconciler) _delete(bucket *bucketv1alpha3.S3Bucket, client s3.Service) (reconcile.Result, error) {
-	bucket.Status.SetConditions(runtimev1alpha1.Deleting(), runtimev1alpha1.ReconcileSuccess())
-	if bucket.Spec.ReclaimPolicy == runtimev1alpha1.ReclaimDelete {
-		if err := client.DeleteBucket(bucket); err != nil {
-			return r.fail(bucket, err)
-		}
+func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
```

Comment:
In `Update`, you should make the `Put` call directly. If they do what's expected from a `PUT`, i.e. update or create if not exists, we shouldn't need to worry about patches or differences etc. We'd call `PUT` in any case and move on. Though this needs manual check, it might return `AlreadyExists` error if it's already there.

Reply:
Okay, I checked that `PutBucketPolicyRequest` doesn't return any error for already existing policy. So we can call it without a check. 
Just that a local change can save an API call. And there will be more API calls for more fields we add, because there is no single consolidated call for S3 bucket properties.